#! /bin/sh /usr/share/dpatch/dpatch-run
## 04_bonus_control_code_in_clients_conf_5.dpatch by Paul Hampson <Paul.Hampson@anu.edu.au>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: A check for a useul 'op' value in rlm_sql's sql.c (part of a segfault fix
## DP: between 1.0.2 and 1.0.3) is reusing a buffer that needed to survive from
## DP: the attribute check until the pair assembly.
## DP: Pulled from CVS: http://www.freeradius.org/cgi-bin/cvsweb.cgi/radiusd/src/modules/rlm_sql/sql.c.diff?r1=1.79.2.1&r2=1.79.2.2

@DPATCH@
diff -u -p -r1.79.2.1 -r1.79.2.2
--- radiusd/src/modules/rlm_sql/sql.c	2005/05/20 09:13:32	1.79.2.1
+++ radiusd/src/modules/rlm_sql/sql.c	2005/06/25 22:45:35	1.79.2.2
@@ -341,6 +341,22 @@ int sql_userparse(VALUE_PAIR ** first_pa
 	}
 
 	/*
+	 *	Verify the 'op' field
+	 */
+	if (row[4] != NULL && row[4][0] != '\0') {
+		ptr = row[4];
+		operator = gettoken(&ptr, buf, sizeof(buf));
+	}
+	if (operator <= T_EOL) {
+		/*
+		 *  Complain about empty or invalid 'op' field
+		 */
+		operator = T_OP_CMP_EQ;
+		radlog(L_ERR, "rlm_sql: The 'op' field for attribute '%s = %s' is NULL, or non-existent.", row[2], row[3]);
+		radlog(L_ERR, "rlm_sql: You MUST FIX THIS if you want the configuration to behave as you expect.");
+	}
+
+	/*
 	 *	The 'Value' field may be empty or NULL
 	 */
 	value = row[3];
@@ -352,7 +368,6 @@ int sql_userparse(VALUE_PAIR ** first_pa
 	   ((row[3][0] == '\'') || (row[3][0] == '`') || (row[3][0] == '"')) &&
 	   (row[3][0] == row[3][strlen(row[3])-1])) {
 
-		value = row[3];
 		token = gettoken(&value, buf, sizeof(buf));
 		switch (token) {
 			/*
@@ -378,22 +393,6 @@ int sql_userparse(VALUE_PAIR ** first_pa
 			value = row[3];
 			break;
 		}
-	}
-
-	/*
-	 *	Verify the 'op' field
-	 */
-	if (row[4] != NULL && row[4][0] != '\0') {
-		ptr = row[4];
-		operator = gettoken(&ptr, buf, sizeof(buf));
-	}
-	if (operator <= T_EOL) {
-		/*
-		 *  Complain about empty or invalid 'op' field
-		 */
-		operator = T_OP_CMP_EQ;
-		radlog(L_ERR, "rlm_sql: The 'op' field for attribute '%s = %s' is NULL, or non-existent.", row[2], row[3]);
-		radlog(L_ERR, "rlm_sql: You MUST FIX THIS if you want the configuration to behave as you expect.");
 	}
 
 	/*
